external-crates/move/crates/move-compiler/src/typing/translate.rs

```diff
@@ -2543,6 +2918,42 @@ fn builtin_call(
     (ret_ty, call)
 }
 
+fn syntax_call_return_ty(
+    context: &mut Context,
+    loc: Loc,
+    m: ModuleIdent,
+    f: FunctionName,
+    fty: ResolvedFunctionType,
+    argloc: Loc,
+    tys: Vec<Type>,
+) -> Type {
+    let ResolvedFunctionType {
+        declared,
+        macro_,
+        ty_args: _,
+        params: parameters,
+        return_,
+    } = fty;
+    check_call_target(context, loc, macro_, macro_, declared, f);
+    let arg_tys = {
+        let msg = || format!("Invalid call of '{}::{}'", &m, &f);
+        let arity = parameters.len();
+        make_arg_types(context, loc, msg, arity, argloc, tys)
+    };
+    assert!(arg_tys.len() == parameters.len());
+    let mut valid = true;
+    for (arg_ty, (_, param_ty)) in arg_tys.into_iter().zip(parameters.clone()) {
+        if let Err(_failure) = subtype_no_report(context, arg_ty, param_ty) {
+            valid = false;
+        }
```

Comment:
Why do we need to check this now?

Reply:
We do the suptying to seed a polymorphic return type. Consider:

```
fun index<T>(&mut S, x: &T): &T { ... }
```

We don't know what the base type is until we have `T` in our hand and do this. The subtyping does that bit.

As for the error, the failure case for dotted expressions is handing an error expression up the chain: if a field fails to resolve, it hands up an error so that the user doesn't get a new error at every step in the path. We mimic that behavior here: if our arguments are wrong, the user already messed up so we're done with trying to resolve this path.

Happy to discuss this more offline, too.

Reply:
Gotcha gotcha, makes sense. But lets maybe add a comment here about whats going on. It doesn't have to be long, just little annotations as the function progresses